File: pb/src/main/java/org/sqlbuilder/pb/foreign/VnAlias.java
package org.sqlbuilder.pb.foreign;

import org.sqlbuilder.common.Insertable;
import org.sqlbuilder.pb.objects.PbWord;
import org.sqlbuilder.pb.objects.RoleSet;

import java.util.Comparator;
import java.util.HashSet;
import java.util.Set;

public class VnAlias extends Alias implements Insertable
{
	public static final Comparator<VnAlias> COMPARATOR = Comparator //
			.comparing(VnAlias::getPbRoleSet) //
			.thenComparing(VnAlias::getPbWord) //
			.thenComparing(VnAlias::getRef) //
			.thenComparing(VnAlias::getPos);

	static public final Set<VnAlias> SET = new HashSet<>();

	// C O N S T R U C T

	public static VnAlias make(final String clazz, final String pos, final RoleSet pbRoleSet, final PbWord pbWord)
	{
		var a = new VnAlias(clazz, pos, pbRoleSet, pbWord);
		SET.add(a);
		return a;
	}

	protected VnAlias(final String clazz, final String pos, final RoleSet pbRoleSet, final PbWord pbWord)
	{
		super(clazz, pos, pbRoleSet, pbWord);
	}
}

File: pb/src/main/java/org/sqlbuilder/pb/foreign/VnAlias.kt
package org.sqlbuilder.pb.foreign

import org.sqlbuilder.common.Insertable
import org.sqlbuilder.pb.objects.RoleSet
import org.sqlbuilder.pb.objects.Word

open class VnAlias protected constructor(clazz: String, pos: String, pbRoleSet: RoleSet, word: Word) : Alias(clazz, pos, pbRoleSet, word), Insertable {
    companion object {

        val COMPARATOR: Comparator<VnAlias> = Comparator
            .comparing<VnAlias, RoleSet> { it.pbRoleSet }
            .thenComparing { it.pbWord }
            .thenComparing { it.ref }
            .thenComparing { it.pos }

        val SET: MutableSet<VnAlias> = HashSet<VnAlias>()

        fun make(clazz: String, pos: String, pbRoleSet: RoleSet, word: Word): VnAlias {
            val a = VnAlias(clazz, pos, pbRoleSet, word)
            SET.add(a)
            return a
        }
    }
}
